lib/cusparse/broadcast.jl

```diff
@@ -295,16 +294,52 @@ function _getindex(arg::Union{CuSparseDeviceMatrixCSR,CuSparseDeviceMatrixCSC},
         @inbounds arg.nzVal[ptr]
     end
 end
+_getindex(arg::CuDeviceArray, I, ptr) = @inbounds arg[I]
 _getindex(arg, I, ptr) = Broadcast._broadcast_getindex(arg, I)
 
-
 ## sparse broadcast implementation
 
 iter_type(::Type{<:CuSparseMatrixCSC}, ::Type{Ti}) where {Ti} = CSCIterator{Ti}
 iter_type(::Type{<:CuSparseMatrixCSR}, ::Type{Ti}) where {Ti} = CSRIterator{Ti}
 iter_type(::Type{<:CuSparseDeviceMatrixCSC}, ::Type{Ti}) where {Ti} = CSCIterator{Ti}
 iter_type(::Type{<:CuSparseDeviceMatrixCSR}, ::Type{Ti}) where {Ti} = CSRIterator{Ti}
 
+function compute_offsets_kernel(::Type{<:CuSparseVector}, first_row::Ti, last_row::Ti, offsets::AbstractVector{Ti}, args...) where Ti
+    row_ix = threadIdx().x + (blockIdx().x - 1i32) * blockDim().x
+    row    = row_ix + first_row - 1i32
+    
+    row > last_row && return
+
+    # TODO load arg.iPtr slices into shared memory
+    row_is_nnz = 0i32
+    for arg in args
+        if arg isa CuSparseDeviceVector
+            for arg_row in arg.iPtr
+                if arg_row == row 
+                    row_is_nnz = 1i32
+                    @inbounds offsets[row] = row
+                end
+                arg_row > row && break
+            end
+        else
+            @inbounds offsets[row] = row
+            row_is_nnz = 1i32
+        end
+        sync_warp()
+    end
+    # count nnz in the warp
+    offset = 16
+    while (offset >= 1)
+        row_is_nnz = shfl_down_sync(FULL_MASK, row_is_nnz, offset)
+        offset = offset >> 1
+    end
+    if threadIdx().x == 1
+        CUDA.@atomic offsets[last_row+1] += row_is_nnz
+    end
+    return
+end
+# TODO: unify CSC/CSR kernels
```

Comment:
Isn't `compute_offsets_kernel` below already unified?

Reply:
Yes lol my rebase missed this comment